Batch changes (and remove `options[:every]`)

zachahn opened the pull request:
```
This removes the `every` option when initializing Filewatcher.

Though backwards incompatible, it's fairly simple to emulate the
previous behavior.

    filewatcher.watch do |changes|
      # to emulate every=true
      changes.each do |filename, event|
        # ...
      end

      # to emulate every=false
      filename, event = changes.first
      # ...
    end
```

The CLI option `--every` remains unchanged.

AlexWayfer commented on lib/filewatcher.rb:
```diff
@@ -37,7 +36,7 @@ def watch(&on_update)
     @on_update = on_update
     @keep_watching = true
-    yield('', '') if @immediate
+    yield({ "" => "" }) if @immediate
```
This change is irrelevant. Also, we're using RuboCop and here is its config in the repository.

zachahn replied:
Hm could you clarify?

The specs originally had this: `it { is_expected.to eq [['', '']] }`

So I updated it to `it { is_expected.to eq [{ "" => "" }] }`, I thought that seemed like the most consistent solution

I also briefly considered yielding an empty hash. But I realized that that would cause unnecessary breaking changes, here's what I was thinking:

```ruby
# currently with immediate: true, every: false
fw.watch do |filename, event|
  # filename = "", event = ""
end

# batching changes, yielding { "" => "" }
fw.watch do |changes|
  changes.first(1).each do |filename, event|
    # filename = "", event = ""
  end
end

# batching changes, yielding {}
fw.watch do |changes|
  changes.first(1).each do |filename, event|
    # never runs. user must handle `changes = {}` with a special case
  end
end

# batching changes, yielding "", ""
fw.watch do |changes|
  # changes = ""
end
```

And oops, I will run Rubocop! (Thank you for fixing the gemspec! I had a hard time getting Rubocop to run the other day, but it works now 💪)

AlexWayfer replied:
Sorry, my bad, I didn't notice `,` and `=>`.

I like Symbol keys so `=>` is like trigger for me. :sweat_smile: 

Yes, I meant `yield '' => ''` would be the coolest variant. And are you sure that RuboCop is OK with curly braces here? Usually it ask to omit them when a Hash is a single argument. I'm pretty laze to check it myself, but I will if you confirm.

AlexWayfer replied:
And we should setup RuboCop CI task… oh, it's already here! And it's green! OK, sorry. I'll check RuboCop rules.

AlexWayfer replied:
I found: https://github.com/rubocop-hq/rubocop/issues/7641

I think it's resolved. Sorry.

AlexWayfer commented on lib/filewatcher/cycles.rb:
```diff
@@ -35,10 +35,7 @@ def watching_cycle
-        changes.each do |filename, event|
-          on_update.call(filename, event)
-        end
+        on_update.call(@changes.dup)
```
Why do you think that `dup` is required?

zachahn replied:
Hm I made the change because of failing tests. Here's an example! (I retried this after rebasing)

```
  9) Filewatcher#watch when there are new files in subfolders is expected to eq [{"/path/to/filewatcher/spec/tmp/new_sub_folder"=>:updated, "/path/to/filewatcher/spec/tmp/new_sub_folder/file.txt"=>:created}]
     Failure/Error:
       expect(processed).to eq [
         { subfolder => :updated, watch_run.filename => :created }
       ]

       expected: [{"path/to/filewatcher/spec/tmp/new_sub_folder"=>:updated, "/path/to/filewatcher/spec/tmp/new_sub_folder/file.txt"=>:created}]
            got: [{}]

       (compared using ==)

       Diff:
       @@ -1,3 +1,2 @@
       -[{"/path/to/filewatcher/spec/tmp/new_sub_folder"=>:updated,
       -  "/path/to/filewatcher/spec/tmp/new_sub_folder/file.txt"=>:created}]
       +[{}]

     # ./spec/filewatcher_spec.rb:163:in `block (4 levels) in <top (required)>'
```

The `.dup` fixed a bunch of errors very similar to the one above, and each of them expected some non-empty hash but received an empty one.

Let me know if you have a better solution for this! I'm guessing that this used to work without the `dup` since the Enumerator was created before `@change.clear`. Now that we aren't iterating over the hash, it seems that it's being emptied before the test/user gets a chance to read from it.

AlexWayfer replied:
I think, for multi-threading things, like we have ones in specs, it's better to do `dup`, yes. Thank you.

AlexWayfer commented on bin/filewatcher:
```diff
@@ -79,25 +79,29 @@ begin
-      rescue SystemExit, Interrupt
-        exit(0)
+  fw.watch do |changes|
+    result = options[:every] ? changes : changes.first(1)
```
`result` of what? I think, we can even re-assign `changes` here.

AlexWayfer replied:
I see you've resolved this, please comment or even press the "Resolve" button.

AlexWayfer commented:
Oh, I see errors like:

> `expected `[{}].empty?` to return true, got false`

I think it's easy to fix.

Let me know when there will be complicate fails.

AlexWayfer commented:
Tests should work better after #116 (the second commit), so I recommend you to update (rebase) your branch.

zachahn commented:
Thanks for the review! I have a few questions, but I'll work on some of the comments you gave and push it up soon 😄

zachahn commented:
Hm it looks like there's a flaky test here. I had the same issue locally, but I didn't look into it much.

Interestingly as I was running it repeatedly, I got a slightly different error.

```
D, [2020-04-20T17:33:49.370980 #67025] DEBUG -- : Thread #70136262500420 processed = []
#<Thread:0x00007f93be1fa708@/path/to/filewatcher/spec/spec_helper.rb:135 run> terminated with exception (report_on_exception is true):
Traceback (most recent call last):
	9: from /path/to/filewatcher/spec/spec_helper.rb:135:in `block in thread_initialize'
	8: from /path/to/filewatcher/spec/spec_helper.rb:140:in `setup_filewatcher'
	7: from /path/to/filewatcher/lib/filewatcher.rb:42:in `watch'
	6: from /path/to/filewatcher/lib/filewatcher/cycles.rb:14:in `main_cycle'
	5: from /path/to/filewatcher/lib/filewatcher/cycles.rb:30:in `watching_cycle'
	4: from /path/to/filewatcher/lib/filewatcher.rb:111:in `filesystem_updated?'
	3: from /path/to/filewatcher/lib/filewatcher.rb:104:in `mtime_snapshot'
	2: from /path/to/filewatcher/lib/filewatcher.rb:104:in `each'
	1: from /path/to/filewatcher/lib/filewatcher.rb:105:in `block in mtime_snapshot'
/path/to/filewatcher/lib/filewatcher.rb:105:in `mtime': No such file or directory @ rb_file_s_mtime - /path/to/filewatcher/spec/tmp/tmp_file.txt (Errno::ENOENT)
```

AlexWayfer commented:
> Interestingly as I was running it repeatedly, I got a slightly different error.

Oh, thank you. I usually see fails like this (from your branch): https://github.com/filewatcher/filewatcher/pull/115/checks?check_run_id=603280959

I'm trying to improve the situation in #119, but… with macOS, JRuby, CLI, files, threads, etc — it's very hard!

Sorry about this. I'll notice about updates.

AlexWayfer commented:
@zachahn hello! Can you please rebase the branch? #119 and many other improvements of specs and CI was merged into `master`.

AlexWayfer commented:
@zachahn sorry, but there are conflicts again… CI should pass faster and more accurate! Please, update the PR if you have a chance. I can try to help if you want.

zachahn commented:
@AlexWayfer No worries! Rebased and pushed

AlexWayfer commented on spec/spec_helper.rb:
```diff
@@ -169,17 +169,23 @@ def thread_initialize
+    changes.each do |filename, event|
+      debug "watch callback: filename = #{filename}, event = #{event}"
+    end
+  end
```
We can try to avoid this additional method with something like:

```ruby
debug "watch callback: changes = #{changes}" # maybe `changes.inspect`, but it will not be better as I know
```

What do you think?

zachahn replied:
Sure, I can change that! I don't have much of an opinion either way, but your suggestion might make it easier to debug :)

AlexWayfer commented:
Pretty nice PR, thank you. I'm waiting for your response about new debug method.

zachahn commented:
@AlexWayfer Updated! I opted to use `changes.inspect`, but you're right, there was no difference in the output

AlexWayfer commented:
@zachahn I see a fail with Ruby 2.5: https://cirrus-ci.com/task/5304049320853504?command=test#L1175

And it seems syntax error for this version, not something phantom. So, can you please fix it?

zachahn commented:
Strange! The only thing I changed was that debugging thing, so I'm inclined to believe it's a flaky test. Did you try rerunning the test? (Here's a diff showing what I changed, but since I rebased I'm not sure how long this link will work <https://github.com/filewatcher/filewatcher/compare/929c5da3..d9f8914c>. And it looks like 929c5da3 passed for Ruby 2.5)

I see that the test expects one hash, but it received two hashes. I'm not sure if this is the best way, but perhaps the best way to proceed would be to update the test to merge all the hashes together with something like `changes.reduce(:merge)`.

I won't have time to look at this for the next few days. I'd appreciate it if you have a chance to fix it! I think I checked the box that lets maintainers push onto this branch. But otherwise I'll take a look at it later this week.

AlexWayfer commented:
> Did you try rerunning the test?

No.

Now I've tried and it passes. Hm…

> I see that the test expects one hash, but it received two hashes. I'm not sure if this is the best way, but perhaps the best way to proceed would be to update the test to merge all the hashes together with something like `changes.reduce(:merge)`.

I've understood the reason: there were two Filewatcher's triggers instead of expected one. I think, it's maybe OK to do `reduce(:merge)`, or even add a sleep. And… `merge` can silent other unusual errors. So, because there is one action, `create` of a file, I suggest to wait more time. Due to this was splitted into two hashes, increase the interval of Filewatcher instance for this test.

> I won't have time to look at this for the next few days. I'd appreciate it if you have a chance to fix it! I think I checked the box that lets maintainers push onto this branch. But otherwise I'll take a look at it later this week.

OK, thank you for your work and messages. I've tried and pushed a commit to this PR, so let's see.

AlexWayfer commented:
Now it looks correct and passing, thank you again for your contributions.

zachahn commented:
Thank you for taking a look and fixing it! Happy to contribute, it was a pleasure :)

AlexWayfer commented:
We forgot to update the README…
